pkg/archive/archive.go

```diff
@@ -225,6 +225,61 @@ func CompressStream(dest io.Writer, compression Compression) (io.WriteCloser, er
 	}
 }
 
+// ReplaceFileTarWrapper converts inputTarStream to a new tar stream
+// while replacing a single file called header.Name with new contents.
+// If the file with header.Name does not exist it is added to the tar stream.
+// TODO: make this into a generic tar conversion function with walkFn argument
+func ReplaceFileTarWrapper(inputTarStream io.ReadCloser, header tar.Header, file []byte) io.ReadCloser {
+	pipeReader, pipeWriter := io.Pipe()
+
+	go func() {
+		tarReader := tar.NewReader(inputTarStream)
+		tarWriter := tar.NewWriter(pipeWriter)
+
+		defer inputTarStream.Close()
+
+		for {
+			hdr, err := tarReader.Next()
+			if err == io.EOF {
+				header.Size = int64(len(file))
+
+				if err := tarWriter.WriteHeader(&header); err != nil {
+					pipeWriter.CloseWithError(err)
+					return
+				}
+
+				content := bytes.NewBuffer(file)
+				if _, err := io.Copy(tarWriter, content); err != nil {
+					pipeWriter.CloseWithError(err)
+					return
+				}
+
+				tarWriter.Close()
+				pipeWriter.Close()
```

Comment:
Not sure how important it is for how this tar stream is used but after reading the tar content there could be remaining data that on the input tar stream, we often pipe that to null to ensure that anything blocking on that read is freed (without needing to rely on the close).